File: csharp/DVSClient/DVSClient/address/lookup/LookupSuggestion.cs
using DVSClient.Server.Address.Lookup;

namespace DVSClient.Address.Lookup
{
    public class LookupSuggestion
    {
        public LookupSuggestionLocality? Locality { get; }
        public LookupSuggestionPostalCode? PostalCode { get; }
        public string? PostalCodeKey { get; }
        public string? LocalityKey { get; }
        public LookupSuggestionWhat3Words? What3Words { get; }

        public LookupSuggestion(RestApiAddressLookupSuggestion response)
        {
            Locality = response.Locality != null ? new LookupSuggestionLocality(response.Locality) : null;            
            PostalCode = response.PostalCode != null ? new LookupSuggestionPostalCode(response.PostalCode) : null;
            PostalCodeKey = response.PostalCodeKey;
            LocalityKey = response.LocalityKey;
            What3Words = response.What3words != null ? new LookupSuggestionWhat3Words(response.What3words) : null;
        }
    }

    public class LookupSuggestionLocality
    {
        public LookupLocalityItem? Region { get; }
        public LookupLocalityItem? SubRegion { get; }
        public LookupLocalityItem? Town { get; }
        public LookupLocalityItem? District { get; }
        public LookupLocalityItem? SubDistrict { get; }

        public LookupSuggestionLocality(RestApiAddressLookupSuggestionLocality response)
        {
            Region = response.Region != null ? new LookupLocalityItem(response.Region) : null;
            SubRegion = response.SubRegion != null ? new LookupLocalityItem(response.SubRegion) : null;
            Town = response.Town != null ? new LookupLocalityItem(response.Town) : null;
            District = response.District != null ? new LookupLocalityItem(response.District) : null;
            SubDistrict = response.SubDistrict != null ? new LookupLocalityItem(response.SubDistrict) : null;
        }
    }

    public class LookupSuggestionPostalCode
    {
        public string? FullName { get; }
        public string? Primary { get; }
        public string? Secondary { get; }

        public LookupSuggestionPostalCode(RestApiAddressLookupSuggestionPostalCode response)
        {
            FullName = response.FullName;
            Primary = response.Primary;
            Secondary = response.Secondary;
        }
    }

    public class LookupSuggestionWhat3Words
    {
        public string? Name { get; }
        public string? Description { get; }

        public LookupSuggestionWhat3Words(RestApiAddressLookupSuggestionW3W response)
        {
            Name = response.Name;
            Description = response.Description;
        }
    }

    
}

File: typescript/src/address/lookup/lookupSuggestion.ts
import { RestApiAddressLookupSuggestion } from "../../server/address/lookup/restApiAddresslookupSuggestion";
import { LookupLocalityItem, restApiResponseToLookupLocalityItem } from "./lookupLocalityItem";


export type LookupSuggestion = {
    locality?: {
        region?: LookupLocalityItem;
        subRegion?: LookupLocalityItem;
        town?: LookupLocalityItem;
        district?: LookupLocalityItem;
        subDistrict?: LookupLocalityItem;
    };
    postalCode?: {
        fullName?: string;
        primary?: string;
        secondary?: string;
    };
    postalCodeKey?: string;
    localityKey?: string;
    what3words?: {
        name?: string;
        description?: string;
    };

}

export function restApiResponseToLookupSuggestion(response: RestApiAddressLookupSuggestion): LookupSuggestion {

    const result: LookupSuggestion = {};
    if (response.locality) {
        result.locality = {
            region: response.locality.region ? restApiResponseToLookupLocalityItem(response.locality.region) : {},
            subRegion: response.locality.sub_region ? restApiResponseToLookupLocalityItem(response.locality.sub_region) : {},
            town: response.locality.town ? restApiResponseToLookupLocalityItem(response.locality.town) : {},
            district: response.locality.district ? restApiResponseToLookupLocalityItem(response.locality.district) : {},
            subDistrict: response.locality.sub_district ? restApiResponseToLookupLocalityItem(response.locality.sub_district) : {}
        }
    }
    if (response.postal_code) {
        result.postalCode = {
            fullName: response.postal_code.full_name,
            primary: response.postal_code.primary,
            secondary: response.postal_code.secondary
        }
    }
    if (response.what3words) {
        result.what3words = {
            name: response.what3words.name,
            description: response.what3words.description
        }
    }

    if (response.postal_code_key) {
        result.postalCodeKey = response.postal_code_key;
    }
    if (response.locality_key) {
        result.localityKey = response.locality_key;
    }

    return result;

}